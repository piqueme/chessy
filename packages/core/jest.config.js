const base = require('../../jest.config.base')
const package = require('./package.json')

module.exports = {
  ...base,
  name: package.name,
  displayName: package.name,
  setupFiles: ['<rootDir>/src/__tests__/testSetup.ts']
}
