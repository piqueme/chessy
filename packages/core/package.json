{
  "name": "@chessy/core",
  "private": true,
  "packageManager": "yarn@3.0.2",
  "devDependencies": {
    "@types/jest": "^27.0.2",
    "@types/node": "^16.9.2",
    "@types/uuid": "^8.3.4"
  },
  "main": "build/cjs/index.js",
  "module": "build/mjs/index.js",
  "exports": {
    ".": {
      "import": "./build/mjs/index.js",
      "require": "./build/cjs/index.js"
    }
  },
  "types": "build/mjs/index.d.ts",
  "files": [
    "build"
  ],
  "scripts": {
    "build": "yarn g:tsc -p tsconfig.json && yarn g:tsc -p tsconfig-cjs.json",
    "clean": "rm -rf ./build && rm tsconfig.tsbuildinfo",
    "dev": "yarn g:tsc -b -w --preserveWatchOutput",
    "test": "yarn g:test --selectProjects @chessy/core"
  },
  "dependencies": {
    "pino": "^7.6.5",
    "uuid": "^8.3.2"
  }
}
