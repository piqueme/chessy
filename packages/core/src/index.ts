export type { Side, Square, Piece, Board } from './board'
export type { Puzzle } from './puzzle'
export type { Move, HistoryMove } from './moves'
export type { PuzzleMasterGame, PlayerGame, History } from './game'

export { default as Log, LoggingScope } from './logger'
export { createMasterGame, createPlayerGame, moveMasterGame, movePlayerGame } from './game'
export { isValidMove, canPromoteFromAssumedValidMove, parseMoveNotation, executeMove } from './moves'
export { getAllSquares, atSquare, readCompressedBoard, readBoard, serializeBoard, serializeCompressedBoard, getEnemySide } from './board'
