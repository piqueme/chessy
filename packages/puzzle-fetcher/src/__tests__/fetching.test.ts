import fs from 'fs'
import path from 'path'
import axios from 'axios'
import * as core from '@chessy/core'
import {
  fetchBlunder,
  parseBlunder,
  fetch
} from '../fetching'
import type {
  Blunder,
} from '../fetching'
import type { Board, Piece, HistoryMove } from '@chessy/core'

jest.mock('axios')
const mockedAxios = axios as jest.Mocked<typeof axios>

const mockBlunder: Blunder = {
  id: 'mock-blunder-id',
  elo: 1449,
  blunderMove: 'c5',
  fenBefore: '8/2p5/1pPp1k1p/p2P1p2/P1Pn1n1P/2n5/6B1/7K w - - 3 49',
  forcedLine: ['bxc5'],
  move_index: 96,
}

describe('fetching blunder', () => {
  test('makes single request through axios', async () => {
    mockedAxios.post.mockResolvedValueOnce({
      data: {
        data: mockBlunder
      }
    })
    const resultBlunder = await fetchBlunder()
    expect(mockedAxios.post).toHaveBeenCalledTimes(1)
    expect(resultBlunder).toEqual(mockBlunder)
  })

  test('rejects with error when API call fails', async () => {
    const errorMessage = 'Network error'
    mockedAxios.post.mockRejectedValueOnce(new Error(errorMessage))
    await expect(
      async () => { await fetchBlunder() }
    ).rejects.toThrow(errorMessage)
  })
})

// NOTE: This test is focused on _unit_ testing parsing, so we mock external
// dependencies on core logic.
describe('parsing blunder', () => {
  const readCompressedBoardSpy = jest.spyOn(core, 'readCompressedBoard')
  const parseMoveNotationSpy = jest.spyOn(core, 'parseMoveNotation')
  const executeMoveSpy = jest.spyOn(core, 'executeMove')

  const mockSideToMove = 'black'
  const getEnemySideSpy = jest.spyOn(core, 'getEnemySide')

  const moveOne: HistoryMove['move'] = { from: [2, 3], to: [1, 3] }
  const moveTwo: HistoryMove['move'] = { from: [1, 3], to: [0, 3] }
  const moveTwoNotation = mockBlunder.forcedLine[0]

  const whiteKing: Piece = { type: 'king', side: 'white' }
  const boardBeforeBlunderMove: Board = [
    [null, null, null, null, null, null, null, null],
    [null, null, null, null, null, null, null, null],
    [null, null, null, whiteKing, null, null, null, null],
    [null, null, null, null, null, null, null, null],
    [null, null, null, null, null, null, null, null],
    [null, null, null, null, null, null, null, null],
    [null, null, null, null, null, null, null, null],
    [null, null, null, null, null, null, null, null]
  ]
  const boardAfterBlunderMove: Board = [
    [null, null, null, null, null, null, null, null],
    [null, null, null, whiteKing, null, null, null, null],
    [null, null, null, null, null, null, null, null],
    [null, null, null, null, null, null, null, null],
    [null, null, null, null, null, null, null, null],
    [null, null, null, null, null, null, null, null],
    [null, null, null, null, null, null, null, null],
    [null, null, null, null, null, null, null, null]
  ]
  const boardAfterAllPuzzleMoves: Board = [
    [null, null, null, whiteKing, null, null, null, null],
    [null, null, null, null, null, null, null, null],
    [null, null, null, null, null, null, null, null],
    [null, null, null, null, null, null, null, null],
    [null, null, null, null, null, null, null, null],
    [null, null, null, null, null, null, null, null],
    [null, null, null, null, null, null, null, null],
    [null, null, null, null, null, null, null, null]
  ]

  beforeEach(() => {
    readCompressedBoardSpy.mockReturnValue(boardBeforeBlunderMove)

    parseMoveNotationSpy
      .mockReturnValueOnce(moveOne)
      .mockReturnValueOnce(moveTwo)

    executeMoveSpy
      .mockReturnValueOnce({
        fullMove: moveOne,
        newBoard: boardAfterBlunderMove
      })
      .mockReturnValueOnce({
        fullMove: moveTwo,
        newBoard: boardAfterAllPuzzleMoves
      })

    getEnemySideSpy.mockReturnValue(mockSideToMove)
  })

  afterAll(() => {
    readCompressedBoardSpy.mockRestore()
    parseMoveNotationSpy.mockRestore()
    executeMoveSpy.mockRestore()
    getEnemySideSpy.mockRestore()
  })

  test('parses blunder move separately first as first output puzzle move', () => {
    parseBlunder(mockBlunder)
    expect(readCompressedBoardSpy).toHaveBeenCalledTimes(1)
    expect(parseMoveNotationSpy).toHaveBeenCalledTimes(2)
    expect(executeMoveSpy).toHaveBeenCalledTimes(2)
  })

  test('outputs puzzle with start board after blunder move and moves excluding blunder move in history', () => {
    const parsedPuzzle = parseBlunder(mockBlunder)
    expect(parsedPuzzle).toEqual({
      startBoard: boardAfterBlunderMove,
      sideToMove: mockSideToMove,
      correctMoves: [{ move: moveTwo, notation: moveTwoNotation }]
    })
  })
})

describe('[INTEGRATION] blunder fetching and parsing pipeline', () => {
  test('fetches and parses puzzles, throwing out fetch/parse failures', async () => {
    const mockDir = path.join(__dirname, 'mocks')
    const readBlunder = (filename: string): Blunder =>
      JSON.parse(fs.readFileSync(path.join(mockDir, filename), 'utf-8'))

    const blunder1 = readBlunder('blunder-1.json')
    const blunder2 = readBlunder('blunder-2.json')
    const brokenBlunder3 = readBlunder('blunder-3-broken.json')
    const blunder4 = readBlunder('blunder-4.json')

    mockedAxios.post.mockResolvedValueOnce({
      data: { data: blunder1 }
    })
    .mockResolvedValueOnce({
      data: { data: blunder2 }
    })
    .mockRejectedValueOnce(new Error('Blunder fetch error'))
    .mockResolvedValueOnce({
      data: { data: brokenBlunder3 }
    })
    .mockResolvedValueOnce({
      data: { data: blunder4 }
    })

    const puzzles = await fetch({
      numPuzzles: 5,
      pause: 100,
      concurrency: 1
    })

    expect(puzzles).toHaveLength(3)
    expect(puzzles).toMatchSnapshot()
  })
})
