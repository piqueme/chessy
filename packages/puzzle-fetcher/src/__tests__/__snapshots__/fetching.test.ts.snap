// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`[INTEGRATION] blunder fetching and parsing pipeline fetches and parses puzzles, throwing out fetch/parse failures 1`] = `
Array [
  Object {
    "_id": "5579ebe8e13823b81c3dfa46",
    "correctMoves": Array [
      Object {
        "move": Object {
          "from": Array [
            3,
            2,
          ],
          "take": Object {
            "piece": Object {
              "_id": "mock-id",
              "side": "white",
              "type": "queen",
            },
            "square": Array [
              6,
              5,
            ],
          },
          "to": Array [
            6,
            5,
          ],
        },
        "notation": "Qxf2",
      },
      Object {
        "move": Object {
          "from": Array [
            7,
            3,
          ],
          "to": Array [
            6,
            3,
          ],
        },
        "notation": "Rd2",
      },
      Object {
        "move": Object {
          "from": Array [
            6,
            5,
          ],
          "to": Array [
            5,
            4,
          ],
        },
        "notation": "Qe3",
      },
    ],
    "sideToMove": "black",
    "startBoard": Array [
      Array [
        null,
        null,
        null,
        null,
        null,
        null,
        Object {
          "_id": "mock-id",
          "side": "black",
          "type": "rook",
        },
        null,
      ],
      Array [
        null,
        null,
        null,
        null,
        Object {
          "_id": "mock-id",
          "side": "black",
          "type": "bishop",
        },
        null,
        Object {
          "_id": "mock-id",
          "side": "black",
          "type": "king",
        },
        null,
      ],
      Array [
        Object {
          "_id": "mock-id",
          "side": "black",
          "type": "pawn",
        },
        null,
        null,
        null,
        null,
        Object {
          "_id": "mock-id",
          "side": "black",
          "type": "pawn",
        },
        null,
        null,
      ],
      Array [
        null,
        Object {
          "_id": "mock-id",
          "side": "black",
          "type": "pawn",
        },
        Object {
          "_id": "mock-id",
          "side": "black",
          "type": "queen",
        },
        Object {
          "_id": "mock-id",
          "side": "white",
          "type": "pawn",
        },
        null,
        Object {
          "_id": "mock-id",
          "side": "black",
          "type": "bishop",
        },
        null,
        null,
      ],
      Array [
        null,
        null,
        null,
        null,
        null,
        null,
        null,
        null,
      ],
      Array [
        null,
        null,
        null,
        null,
        null,
        null,
        null,
        null,
      ],
      Array [
        Object {
          "_id": "mock-id",
          "side": "white",
          "type": "pawn",
        },
        Object {
          "_id": "mock-id",
          "side": "white",
          "type": "pawn",
        },
        Object {
          "_id": "mock-id",
          "side": "white",
          "type": "pawn",
        },
        null,
        null,
        Object {
          "_id": "mock-id",
          "side": "white",
          "type": "queen",
        },
        Object {
          "_id": "mock-id",
          "side": "white",
          "type": "pawn",
        },
        null,
      ],
      Array [
        null,
        null,
        Object {
          "_id": "mock-id",
          "side": "white",
          "type": "king",
        },
        Object {
          "_id": "mock-id",
          "side": "white",
          "type": "rook",
        },
        null,
        null,
        null,
        Object {
          "_id": "mock-id",
          "side": "white",
          "type": "rook",
        },
      ],
    ],
  },
  Object {
    "_id": "557ec911e13823b832e33b35",
    "correctMoves": Array [
      Object {
        "move": Object {
          "from": Array [
            4,
            5,
          ],
          "take": Object {
            "piece": Object {
              "_id": "mock-id",
              "side": "black",
              "type": "knight",
            },
            "square": Array [
              5,
              5,
            ],
          },
          "to": Array [
            5,
            5,
          ],
        },
        "notation": "Rxf3",
      },
      Object {
        "move": Object {
          "from": Array [
            1,
            1,
          ],
          "take": Object {
            "piece": Object {
              "_id": "mock-id",
              "side": "white",
              "type": "rook",
            },
            "square": Array [
              5,
              5,
            ],
          },
          "to": Array [
            5,
            5,
          ],
        },
        "notation": "Bxf3",
      },
      Object {
        "move": Object {
          "from": Array [
            5,
            4,
          ],
          "to": Array [
            4,
            5,
          ],
        },
        "notation": "Bf4",
      },
    ],
    "sideToMove": "white",
    "startBoard": Array [
      Array [
        null,
        null,
        Object {
          "_id": "mock-id",
          "side": "black",
          "type": "king",
        },
        Object {
          "_id": "mock-id",
          "side": "black",
          "type": "rook",
        },
        null,
        null,
        Object {
          "_id": "mock-id",
          "side": "black",
          "type": "rook",
        },
        null,
      ],
      Array [
        null,
        Object {
          "_id": "mock-id",
          "side": "black",
          "type": "bishop",
        },
        Object {
          "_id": "mock-id",
          "side": "black",
          "type": "queen",
        },
        Object {
          "_id": "mock-id",
          "side": "black",
          "type": "knight",
        },
        null,
        null,
        null,
        null,
      ],
      Array [
        Object {
          "_id": "mock-id",
          "side": "black",
          "type": "pawn",
        },
        null,
        null,
        null,
        null,
        null,
        null,
        Object {
          "_id": "mock-id",
          "side": "black",
          "type": "pawn",
        },
      ],
      Array [
        null,
        null,
        null,
        null,
        null,
        Object {
          "_id": "mock-id",
          "side": "white",
          "type": "bishop",
        },
        null,
        Object {
          "_id": "mock-id",
          "side": "white",
          "type": "pawn",
        },
      ],
      Array [
        Object {
          "_id": "mock-id",
          "side": "white",
          "type": "pawn",
        },
        null,
        Object {
          "_id": "mock-id",
          "side": "black",
          "type": "pawn",
        },
        Object {
          "_id": "mock-id",
          "side": "white",
          "type": "queen",
        },
        null,
        Object {
          "_id": "mock-id",
          "side": "white",
          "type": "rook",
        },
        Object {
          "_id": "mock-id",
          "side": "white",
          "type": "pawn",
        },
        null,
      ],
      Array [
        null,
        Object {
          "_id": "mock-id",
          "side": "black",
          "type": "pawn",
        },
        Object {
          "_id": "mock-id",
          "side": "white",
          "type": "pawn",
        },
        null,
        Object {
          "_id": "mock-id",
          "side": "white",
          "type": "bishop",
        },
        Object {
          "_id": "mock-id",
          "side": "black",
          "type": "knight",
        },
        Object {
          "_id": "mock-id",
          "side": "white",
          "type": "knight",
        },
        null,
      ],
      Array [
        null,
        Object {
          "_id": "mock-id",
          "side": "white",
          "type": "pawn",
        },
        null,
        null,
        null,
        null,
        null,
        null,
      ],
      Array [
        Object {
          "_id": "mock-id",
          "side": "white",
          "type": "rook",
        },
        null,
        null,
        null,
        null,
        null,
        Object {
          "_id": "mock-id",
          "side": "white",
          "type": "king",
        },
        null,
      ],
    ],
  },
  Object {
    "_id": "557a8640e13823b81d8200d7",
    "correctMoves": Array [
      Object {
        "move": Object {
          "from": Array [
            2,
            0,
          ],
          "to": Array [
            2,
            5,
          ],
        },
        "notation": "Rf6+",
      },
      Object {
        "move": Object {
          "from": Array [
            1,
            5,
          ],
          "to": Array [
            0,
            6,
          ],
        },
        "notation": "Kg8",
      },
      Object {
        "move": Object {
          "from": Array [
            6,
            4,
          ],
          "take": Object {
            "piece": Object {
              "_id": "mock-id",
              "side": "black",
              "type": "rook",
            },
            "square": Array [
              0,
              4,
            ],
          },
          "to": Array [
            0,
            4,
          ],
        },
        "notation": "Rxe8+",
      },
    ],
    "sideToMove": "white",
    "startBoard": Array [
      Array [
        null,
        null,
        null,
        null,
        Object {
          "_id": "mock-id",
          "side": "black",
          "type": "rook",
        },
        null,
        null,
        null,
      ],
      Array [
        null,
        Object {
          "_id": "mock-id",
          "side": "black",
          "type": "rook",
        },
        null,
        null,
        null,
        Object {
          "_id": "mock-id",
          "side": "black",
          "type": "king",
        },
        null,
        Object {
          "_id": "mock-id",
          "side": "black",
          "type": "pawn",
        },
      ],
      Array [
        Object {
          "_id": "mock-id",
          "side": "white",
          "type": "rook",
        },
        null,
        null,
        null,
        null,
        null,
        null,
        null,
      ],
      Array [
        null,
        Object {
          "_id": "mock-id",
          "side": "black",
          "type": "pawn",
        },
        null,
        Object {
          "_id": "mock-id",
          "side": "black",
          "type": "bishop",
        },
        null,
        Object {
          "_id": "mock-id",
          "side": "white",
          "type": "pawn",
        },
        null,
        null,
      ],
      Array [
        Object {
          "_id": "mock-id",
          "side": "black",
          "type": "pawn",
        },
        null,
        null,
        Object {
          "_id": "mock-id",
          "side": "white",
          "type": "bishop",
        },
        null,
        Object {
          "_id": "mock-id",
          "side": "white",
          "type": "king",
        },
        null,
        Object {
          "_id": "mock-id",
          "side": "white",
          "type": "pawn",
        },
      ],
      Array [
        null,
        null,
        Object {
          "_id": "mock-id",
          "side": "white",
          "type": "pawn",
        },
        null,
        null,
        null,
        null,
        null,
      ],
      Array [
        null,
        null,
        Object {
          "_id": "mock-id",
          "side": "white",
          "type": "pawn",
        },
        null,
        Object {
          "_id": "mock-id",
          "side": "white",
          "type": "rook",
        },
        null,
        null,
        null,
      ],
      Array [
        null,
        null,
        null,
        null,
        null,
        null,
        null,
        null,
      ],
    ],
  },
]
`;
